libminifi/include/core/logging/LoggingInterceptor.h

```diff
@@ -0,0 +1,117 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef LIBMINIFI_INCLUDE_CORE_LOGGING_LOGGINGINTERCEPTOR_H_
+#define LIBMINIFI_INCLUDE_CORE_LOGGING_LOGGINGINTERCEPTOR_H_
+
+#include <vector>
+#include <mutex>
+#include <algorithm>
+#include "core/expect.h"
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace core {
+namespace logging {
+
+class Interceptor {
+ public:
+  void intercept_error(const std::string &msg) {
+    std::lock_guard<std::mutex> lock(guard_);
+    errors_.emplace_back(msg);
+  }
+
+  const std::vector<std::string> &getErrors() const {
+    return errors_;
+  }
+
+ protected:
+  std::mutex guard_;
+  std::vector<std::string> errors_;
+};
+
+class LoggingInterceptors {
+ public:
+  static LoggingInterceptors *getInstance() {
+    static LoggingInterceptors interceptors;
+    return &interceptors;
+  }
+
+  void addInterceptor(Interceptor * interceptor) {
+    std::lock_guard<std::mutex> lock(guard_);
+    interceptors_.emplace_back(interceptor);
+  }
+
+  void removeInterceptor(const Interceptor * const interceptor) {
+    std::lock_guard<std::mutex> lock(guard_);
+    interceptors_.erase(std::remove(interceptors_.begin(), interceptors_.end(), interceptor), interceptors_.end());
+  }
+
+  void intercept_error(const std::string &message) {
+    std::lock_guard<std::mutex> lock(guard_);
+    if (UNLIKELY(!interceptors_.empty())) {
+      for (auto &interceptor : interceptors_) {
+        interceptor->intercept_error(message);
+      }
+    }
+  }
+
+  bool empty() const{
+    std::lock_guard<std::mutex> lock(guard_);
+    return interceptors_.empty();
+  }
+
+ private:
+  LoggingInterceptors() {
+
+  }
+
+  mutable std::mutex guard_;
+
+  std::vector<Interceptor*> interceptors_;
+};
+
+/**
+ * Logging interceptor gives us a way to temporarily retrieve certain log levels
+ * and provide them elsewhere. Currently this is only used for error.
+ */
+class LoggingInterceptor : public Interceptor {
+
+ public:
+
+  LoggingInterceptor() {
+    LoggingInterceptors::getInstance()->addInterceptor(this);
+  }
+
+  ~LoggingInterceptor() {
+    std::lock_guard<std::mutex> lock(guard_);
```

Comment:
This is a possible deadlock. 
LoggingInterceptor is being destructed while LoggingInterceptors::intercept_error is being executed. 
Both of these grab's it's own lock before trying to call a member of the other to lock that, too.

Reply:
Yah that's a bad design, will fix. Though in thinking about it the logging interceptor is not what I want to do. So I will likely remove it entirely.  It should be more purposeful.n

Reply:
Errors should be captured mode purposefully, rather n